vowpalwabbit/interactions.cc

```diff
@@ -227,197 +290,101 @@ void sort_and_filter_duplicate_interactions(
-
-          sum_feat_sq_in_inter *= results[order_of_inter - 1];  // will be explained in http://bit.ly/1Hk9JX1
-
-          // let's calculate  the number of a new features
-
-          // if number of features is less than  order of interaction then go to the next interaction
-          // as you can't make simple combination of interaction 'aaa' if a contains < 3 features.
-          // unless one of them has value != 1. and we are counting them.
-          const size_t ft_size = fs.size();
-          if (cnt_ft_value_non_1 == 0 && ft_size < order_of_inter)
-          {
-            num_features_in_inter = 0;
-            break;
-          }
-
-          size_t n;
-          if (cnt_ft_value_non_1 == 0)  // number of generated simple combinations is C(n,k)
-          {
-            n = static_cast<size_t>(
-                VW::math::choose(static_cast<int64_t>(ft_size), static_cast<int64_t>(order_of_inter)));
-          }
-          else
-          {
-            n = 0;
-            for (size_t l = 0; l <= order_of_inter; ++l)
-            {
-              // C(l+m-1, l) * C(n-m, k-l)
-              size_t num = (l == 0) ? 1 : static_cast<size_t>(VW::math::choose(l + cnt_ft_value_non_1 - 1, l));
+    indices[0] += 1;
+  }
 
-              if (ft_size - cnt_ft_value_non_1 >= order_of_inter - l)
-                num *= static_cast<size_t>(VW::math::choose(ft_size - cnt_ft_value_non_1, order_of_inter - l));
-              else
-                num = 0;
+  return result;
+}
 
-              n += num;
-            }
+std::vector<std::vector<namespace_index>> generate_permutations_with_repetition(
+    const std::set<namespace_index>& namespaces, size_t num_to_pick)
+{
+  std::vector<std::vector<namespace_index>> result;
+  result.reserve(VW::math::number_of_permutations_with_repetition(namespaces.size(), num_to_pick));
 
-          }  // details on http://bit.ly/1Hk9JX1
+  std::vector<size_t> one_based_chosen_indices(num_to_pick, 0);
+  for (size_t i = 0; i < num_to_pick - 1; i++) { one_based_chosen_indices[i] = 1; }
+  one_based_chosen_indices[num_to_pick - 1] = 0;
 
-          num_features_in_inter *= n;
+  size_t number_of_namespaces = namespaces.size();
+  size_t next_index = num_to_pick;
 
-          ns += order_of_inter - 1;  // jump over whole block
-        }
+  while (true)
+  {
+    if (one_based_chosen_indices[next_index - 1] == number_of_namespaces)
+    {
+      next_index--;
+      if (next_index == 0) { break; }
+    }
+    else
+    {
+      one_based_chosen_indices[next_index - 1]++;
+      while (next_index < num_to_pick)
+      {
+        next_index++;
+        one_based_chosen_indices[next_index - 1] = 1;
       }
 
-      if (num_features_in_inter == 0) continue;  // signal that values should be ignored (as default value is 1)
-
-      new_features_cnt += num_features_in_inter;
-      new_features_value += sum_feat_sq_in_inter;
+      result.push_back(indices_to_values_one_based(one_based_chosen_indices, namespaces));
```

Comment:
emplace_back? move? does it matter here?

Reply:
`emplace_back` probably helps, `move` wouldn't do anything since returning from a function is already an rvalue